packages/test-runner-module-mocking/README.md

```diff
@@ -0,0 +1,145 @@
+  const hour = getCurrentHour();
+  if (hour < 6 || hour > 18) {
+    return 'night';
+  }
+  return 'day';
+}
+```
+
+```js
+// test/getTimeOfDay.test.js
+import { importMockable } from '@web/test-runner-module-mocking';
+
+const timeLibrary = await importMockable('time-library');
+const { getTimeOfDay } = await import('../src/getTimeOfDay.js');
+
+describe('getTimeOfDay', () => {
+    it('returns night at 2', () => {
+        timeLibrary.getCurrentHour = () => 2;
+        const result = getTimeOfDay();
+        if (result !== 'night') {
+            throw;
+        }
+    });
+});
+```
+
+### Extended test scenario
+
+This scenario showcases how to use `@web/test-runner-module-mocking` together with `chai` and `sinon`.
+
+```js
+// test/getTimeOfDay.test.js
+import { stub } from 'sinon';
+import { expect } from 'chai';
+import { importMockable } from '@web/test-runner-module-mocking';
+
+const timeLibrary = await importMockable('time-library');
+const { getTimeOfDay } = await import('../src/getTimeOfDay.js');
+
+describe('getTimeOfDay', () => {
+  it('returns night at 2', () => {
+    const stubGetCurrentHour = stub(timeLibrary, 'getCurrentHour').returns(2);
+    try {
+      const result = getTimeOfDay();
+      expect(result).to.equal('night');
+    } finally {
+      stubGetCurrentHour.restore();
+    }
+  });
+  it('returns day at 14', () => {
+    const stubGetCurrentHour = stub(timeLibrary, 'getCurrentHour').returns(14);
+    try {
+      const result = getTimeOfDay();
+      expect(result).to.equal('day');
+    } finally {
+      stubGetCurrentHour.restore();
+    }
+  });
+});
+```
+
+## Caveats
+
+### Order of imports
+
+In order to intercept a module, the module should not be referenced yet. (Once a module is loaded, the module loader also loads all its dependent modules) This means the module containing the system under test should be loaded _after_ the module interception. As interception is achieved using a function, this also means the system under test should _always_ be loaded using a dynamic import in order to be done after the interception.
+
+```js
+import { importMockable } from '@web/test-runner-module-mocking';
+
+// First, intercept
+const externalLibrary = await importMockable('external-library');
+
+// Second, load the system under test.
+const systemUnderTest = await import('../../src/system-under-test.js');
+
+// Run tests
+...
+```
+
+### Types of module specifiers
+
+Currently, two types of module specifiers are supported: bare modules and server relative modules.
+
+```javascript
+// bare module, located in `node_modules`
+await importMockable('external-library');
+
+// server relative module
+await importMockable('/src/library-to-intercept.js');
+```
+
+In order to use regular relative references, `import.meta.resolve()` and `new URL().pathname` can be used.
+
+```javascript
+// relative module
+await importMockable(new URL(import.meta.resolve('../src/library-to-intercept.js')).pathname);
```

Comment:
I think for transparency it's good to mention here that other ideas were considered, because the approach with explicit dynamic import is causing 2 issues
- you need to change some static imports to dynamic ones to ensure that the mockable module is not loaded by your component/app code earlier than the test makes it mockable (all such imports that potentially load the same module should be rewritten from static to dynamic to be loaded after the `importMockable` call)
- the relative imports are quite painful which is also seen here

We discussed the alternative proposal which could solve both issues: put the mockable import on top of the test, use static import for it and a special import attribute, e.g.:

```
// top of the test file
import json from "../src/library-to-intercept.js" with { type: "module-mockable" };

// all other imports

// test code
```

this solve the issue with dynamic imports, as well as makes it really easy to make anything mockable without running into the import ordering issue which is not easy to debug

but it can also help to solve the relative import issue (basically improve the DX) by instrumenting the code dynamically based on this special `type: "module-mockable"`, similar to what `importMockable` does internally, e.g. rewriting the code to include a query param in the import with the relative path

since this is quite a magic, for now we decided to go with the current proposal

I think it would be nice to explore the alternative proposal in the future after gathering feedback from users about the current one

Reply:
Fair point! I added a section explaining this. 😃